### PR TITLE
fix(tests): give single-card and multi-card legs distinct junit-xml

### PR DESCRIPTION
**Pull request by HarikrishnanBalagopal**
## What

`make tests` fans out into two `run_test.sh` invocations, `tests-single-card`
then `tests-multi-card`, and passes both the same `PYTEST_ARGS`. When
`PYTEST_ARGS` carries `--junit-xml=FILE`, both legs write the same `FILE`, and
the second leg erases the first. This gives each leg its own report file.

## The defect

`run_test.sh` writes the caller's `--junit-xml` destination once per invocation
with a truncating merge (it merges its own shard files into that path and
overwrites whatever was there). The `tests:` target runs it twice with one
shared `--junit-xml`:

```make
$(MAKE) tests-single-card ... PYTEST_ARGS="$(PYTEST_ARGS)"
$(MAKE) tests-multi-card  ... PYTEST_ARGS="$(PYTEST_ARGS)"
```

So `make tests TEST_TYPE=trunk PYTEST_ARGS="... --junit-xml=trunk.xml"` runs the
single-card leg (writes `trunk.xml`), then the multi-card leg (writes
`trunk.xml` again, truncating). Only the multi-card results survive in the
report. If the multi-card leg produces an empty report, the single-card results
are lost and the final XML shows 0 tests even though the single-card leg passed.

This only affects the `make tests` composite path. The GHA test matrix is not
affected because it calls `run_test.sh` directly, once per matrix job, each with
its own `junit_xml_path`.

## Fix

`Makefile`, `tests:` fan-out only. Rewrite `PYTEST_ARGS` per leg when the caller
passed `--junit-xml=FILE`:

- single-card leg keeps `FILE` (existing consumers see the same name),
- multi-card leg writes a `<stem>-multi-card.xml` sibling.

```make
_single_args='$(PYTEST_ARGS)'; \
_multi_args="$$(printf '%s' '$(PYTEST_ARGS)' | sed -E 's#(--junit-xml[= ]+[^ ]*)\.xml#\1-multi-card.xml#')"; \
```

CI ingesters already glob `*.xml` (the `TEST_TYPE=perf` comment in this same
target notes `ingest_xml.py globs *.xml non-recursively`), so both files are
collected as separate rows with no ingester or schema change. This matches the
approach in #3843, which preserves per-attempt reports as additional `*.xml`
files rather than a single fixed path.

## Verified

`make -n` parses; the `sed` transform was checked directly across forms:

| input | output |
|---|---|
| `-v --skip-slow --junit-xml=/x/trunk.xml` | `... --junit-xml=/x/trunk-multi-card.xml` |
| `--junit-xml /tmp/trunk.xml --more` | `--junit-xml /tmp/trunk-multi-card.xml --more` |
| `-v --skip-slow` (no junit) | unchanged |
| `--junit-xml=/tmp/report` (no .xml) | unchanged |

The transform touches only the junit path, so multi-word pytest args (for
example a `-m` marker expression) pass through untouched.

## Scope / notes

- This touches shared test infrastructure that every `make tests` caller uses;
  please review rather than self-merge.
- Related and complementary: #4105 fixes a separate defect in
  `tests/distributed/split_output.sh` (an unquoted `$@` that word-split the
  `-m` marker so distributed shards collected 0 tests). Both are needed for a
  correct `make tests TEST_TYPE=trunk` report on a two-card host: #4105 makes
  the distributed leg actually collect tests, and this PR keeps its results
  from clobbering the single-card leg's report.

### CODE REVIEW

**Review by anubhavjana — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -91,10 +91,21 @@ ifeq ($(TEST_TYPE),perf)
...
+# see the same name), and the multi-card leg writes a "<stem>-multi-card.xml"
+# sibling. CI ingesters glob *.xml, so both files are collected. With no
+# --junit-xml the sed is a no-op and both legs run with PYTEST_ARGS unchanged.
 else ifneq ($(wildcard $(TEST_CONFIGS)/.),)
 	@rc=0; \
-	$(MAKE) tests-single-card TEST_TYPE="$(TEST_TYPE)" TEST_CONFIGS="$(TEST_CONFIGS)" PYTEST_ARGS="$(PYTEST_ARGS)" || rc=1; \
-	$(MAKE) tests-multi-card TEST_TYPE="$(TEST_TYPE)" PYTEST_ARGS="$(PYTEST_ARGS)" || rc=1; \
+	_single_args='$(PYTEST_ARGS)'; \
+	_multi_args="$$(printf '%s' '$(PYTEST_ARGS)' | sed -E 's#(--junit-xml[= ]+[^ ]*)\.xml#\1-multi-card.xml#')"; \
+	$(MAKE) tests-single-card TEST_TYPE="$(TEST_TYPE)" TEST_CONFIGS="$(TEST_CONFIGS)" PYTEST_ARGS="$$_single_args" || rc=1; \
```
Comment (anubhavjana) [inline Makefile:107]:
```bash
$(MAKE) tests-single-card TEST_TYPE="$(TEST_TYPE)" TEST_CONFIGS="$(TEST_CONFIGS)" PYTEST_ARGS="$$_single_args" || rc=1; \
$(MAKE) tests-multi-card TEST_TYPE="$(TEST_TYPE)" PYTEST_ARGS="$$_multi_args" || rc=1; \
```
looks like deactivated lines, can you please check

**Review by anubhavjana — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (HarikrishnanBalagopal) [reply Makefile:107]: Nothing is deactivated: those two `$(MAKE)` lines are live recipe lines. GitHub's diff highlighter greys backslash-continued shell inside a Make recipe, so they only look inactive.

The whole `else ifneq` branch is one shell command joined by `\`:

```make
@rc=0; \
_single_args='$(PYTEST_ARGS)'; \
_multi_args="$$(printf '%s' '$(PYTEST_ARGS)' | sed -E 's#(--junit-xml[= ]+[^ ]*)\.xml#\1-multi-card.xml#')"; \
$(MAKE) tests-single-card TEST_TYPE="$(TEST_TYPE)" TEST_CONFIGS="$(TEST_CONFIGS)" PYTEST_ARGS="$$_single_args" || rc=1; \
$(MAKE) tests-multi-card TEST_TYPE="$(TEST_TYPE)" PYTEST_ARGS="$$_multi_args" || rc=1; \
exit $$rc
```

The `|| rc=1` on each leg plus the final `exit $$rc` is deliberate: a recipe reports failure through its last command's status, so without capturing `rc` a passing multi-card leg after a failing single-card leg would make `make tests` exit 0 and hide the failure. This keeps both legs running and still fails the target if either leg fails.

You can confirm it runs by expanding it: `make -n tests TEST_TYPE=trunk PYTEST_ARGS="-v --junit-xml=/tmp/t.xml"` prints both `make tests-single-card ... --junit-xml=/tmp/t.xml` and `make tests-multi-card ... --junit-xml=/tmp/t-multi-card.xml`.

**Comment by HarikrishnanBalagopal** [comment]
Good question, and both are true at once: the multi-card leg **did** win the overwrite, but what it wrote was an **empty** report, so the final file shows 0 collected rather than the multi-card results.

Two defects compound here. This PR (#4106) fixes only the overwrite; #4105 fixes the collection. Walking Gnanesh's run:

1. Single-card leg runs first, collects the real tests, and merges them into `integration.xml`:
   ```
   line 90741: Merged 138 XML shard(s) -> /home/senuser/xml_reports/integration.xml
   ```
2. Multi-card (distributed) leg runs second. Because of the #4105 defect (unquoted `$@` in `split_output.sh` word-split `-m 'not slow__plat_s390x'`), every distributed shard collected 0 tests (`no tests ran`, the cluster around lines 69347-90739). `run_test.sh`'s shard merger writes the caller's `--junit-xml` path **unconditionally** every invocation, even when the shards are empty:
   ```python
   Path(out_path).write_text(merged)   # merged = <testsuite tests="0"> when all shards collected 0
   ```
   so it overwrites `integration.xml` with an empty testsuite:
   ```
   line 92236: Merged 10 XML shard(s) -> /home/senuser/xml_reports/integration.xml
   ```
3. Ingest reads the final file: `tests=0 passed=0 ... None of the test reports contained any result` (lines 92364, 92387).

So the multi-card leg *did* override the single-card leg (exactly the overwrite this PR addresses); it just overrode with `tests=0` because collection was broken by the separate #4105 bug. "Not collected at all" describes the distributed leg specifically, not the whole run: the single-card leg collected ~2182 passes, and they were the thing that got clobbered.

That is why both PRs are needed for a correct two-card `trunk` report:
- **#4105** makes the distributed leg actually collect its tests, and
- **#4106** gives each leg its own junit file (`<stem>.xml` and `<stem>-multi-card.xml`) so neither overwrites the other and CI globs both.

With #4105 alone, the multi-card leg would collect real tests but still overwrite the single-card file. With #4106 alone (this PR), the two files stay separate but the multi-card one would still be empty until #4105 lands.

**Comment by HarikrishnanBalagopal** [comment]
Fair question, and it is worth spelling out because it is the natural next doubt. The short answer: the ClickHouse results come from a code path that never hits this overwrite, and the run that *did* hit it pushed zero rows.

**The affected run pushed 0 rows.** Gnanesh's two-card `trunk` run ingested the clobbered `integration.xml` and inserted exactly nothing:

```
Connecting to ClickHouse at spyre-dashboard-dev-clickhouse...
  run_id=6a9f9a56-...  tests=0  passed=0  failed=0  xpass=0  xfail=0  skipped=0
  Inserted 0 test cases + 0 properties
```

So this run is not a counter-example: it is the case where the overwrite wiped everything, and the dashboard got nothing from it. That is the whole reason the report looked empty.

**Where the dashboard's results actually come from: GHA, which never shares one `--junit-xml` across two legs.** The normal PR/nightly torch-spyre CI does not call `make tests`. It runs each suite as its own matrix job through `run_test.sh` with its own `junit_xml_path`, and ingests per job (`.github/actions/run-test-suite/action.yml`, `_test_matrix.yaml`). Crucially, `distributed_tests` is excluded from the single-card matrix and run by a separate multi-card job (`_test_matrix.yaml`: "distributed_tests configs carry the trunk label too but are exclusively run by the test_multi_spyre job below ... so [the single-card matrix] doesn't also schedule them"). Single-card and multi-card therefore never write the same file in GHA, so there is no overwrite and every suite ingests normally.

So both are true at once:
- ClickHouse has data, because GHA ingests per job and is unaffected.
- The `make tests TEST_TYPE=trunk PYTEST_ARGS="... --junit-xml=X"` path (what Gnanesh ran, and now also what the `main` Makefile does since the `tests:` target fans out to both legs) loses the single-card results to the empty multi-card overwrite. That path is what #4106 fixes.

**Comment by HarikrishnanBalagopal** [comment]
Ran a manual before/after on an x86 host with 2 Spyre cards, stock vs this
branch, everything else identical. Both fixes reproduce and hold.

## #4105: `split_output.sh` `$@` -> `"$@"`

`make tests-multi-card TEST_TYPE=integration PYTEST_ARGS="-v --skip-slow --junit-xml=mc.xml"`
(10 distributed configs, `AIU_WORLD_SIZE=2`). The `# Running:` line is
`split_output.sh` echoing the pytest command it received; both sides get the
same `-m not slow__plat_x86_64`.

Stock (`$@`, word-splits the marker):
```
# Running: python3 -u -m pytest test_broadcast__oot_wrapper.py -v -m not slow__plat_x86_64 --junit-xml=mc__shard_5.xml
collecting ... collected 0 items
============================ no tests ran in 0.00s =============================
ERROR: file or directory not found: slow__plat_x86_64
```

Fixed (`"$@"`, marker stays one argument):
```
# Running: python3 -u -m pytest test_broadcast__oot_wrapper.py -v -m not slow__plat_x86_64 --junit-xml=mc__shard_5.xml
collecting ... collected 13 items
```

Across the 10 distributed configs:

| signal | stock `$@` | fixed `"$@"` |
|---|---|---|
| shards `collected 0 items` | 18 | 2 (configs with no non-slow tests) |
| shards `file or directory not found: slow__plat` | 18 | 0 |
| shards `collected N>0 items` | 0 | 13 (10,13,17,3,4,5 items) |

Stock final merged report `mc.xml` = 160 bytes, `tests="0"` (empty testsuite).

## #4106: Makefile `tests:` fan-out, per-leg junit path

Stock `make tests TEST_TYPE=integration PYTEST_ARGS="... --junit-xml=integration.xml"`
writes both legs to the same path:
```
Merged 47 XML shard(s) -> integration.xml    (single-card leg)
Merged 10 XML shard(s) -> integration.xml    (multi-card leg overwrites)
```
Final `integration.xml` = 160 bytes, `tests="0"`. The single-card 47-shard merge
is destroyed by the (empty) 10-shard distributed merge.

With this branch the fan-out rewrites the multi-card leg's `--junit-xml` to a
`<stem>-multi-card.xml` sibling, so the full run produces two files and keeps
both legs:
```
Merged 47 XML shard(s) -> integration.xml             (single-card leg)
Merged 10 XML shard(s) -> integration-multi-card.xml  (multi-card leg, separate file)

integration.xml            = 808830 bytes, 2540 <testcase> elements
integration-multi-card.xml =  26399 bytes,   68 <testcase> elements, 0 failures/0 errors
```

A small side-by-side with a 1-config single-card leg makes the clobber vs
no-clobber obvious:

- fixed: `report.xml` (8302 bytes, tests="2", single-card **preserved**) +
  `report-multi-card.xml` (separate file).
- stock: one `report.xml` = 160 bytes, tests="0". The single-card result is gone.

## Notes

- On one fixed run the distributed leg collected and started, then hit an
  unrelated flaky-card VFIO error (device busy) after collection. Not the marker
  bug: there is no `file or directory not found` on the fixed side at all; the
  fix changed the failure mode from "0 collected (code bug)" to "collected and
  ran (then a card was busy)".
- Pre-existing and orthogonal to both PRs: the merge helper leaves the outer
  `<testsuites tests="0">` roll-up attribute unrecomputed, so the real counts are
  the individual `<testcase>` elements (2540 and 68 above), not the header. Same
  on any run; worth a separate look if any ingester reads that attribute.

**Review by anubhavjana — APPROVED** [review]
